clean up themes (#2)
bmasaiti/ramblings: content/posts/learning-go-as-java-developer/index.md

# Learning Go as a Java Developer

> There is no need to introduce go lang much, by now its popular enough, However for anyone that's like me a month ago I will shorten and summarise here what Go is about , and why I chose to learn it and how I am doing just that. 

Reviewer: Nit: Go is generally referred to just as "Go", "golang" is used in internet searches, etc. I would just use "Go" here.
```suggestion
There is no need to introduce Go, by now it's popular enough, however for anyone that's like me a month ago I will shorten and summarise here what Go is about, why I chose to learn it, and how I am doing just that. 
```

> Go is a programming language ,statically typed and compiled. I could not say Go is a functional language , even though it supports functional programming, and Go is definitely not an object oriented language even though it suppots composition , dependency injection and interfaces.

Reviewer: Some style suggestions:
```suggestion
Go is a programming language, statically typed and compiled. I could not say Go is a functional language, even though it supports functional programming, and Go is definitely not an object oriented language even though it supports composition, dependency injection and interfaces.
```

Also, as surprising as it may sound, the Go authors definitely describe Go as object-oriented, but with one major difference from other object oriented languages - no inheritance (as you noticed). If you want this to remain your opinion on the language, that's totally fine, just wanted to add some context in case you were unsure.

Author: Thanks for this , went back and found the the tech talk by Rob Pike explaining this.

> For the most part of my career I have been a java developer , fell in love way back in college , Hell I even gatekeeped teaching it when I was a university lecturer(story for another day), I understood and still believe in what Java stands for and the stability and confidence it brings not forgetting the pervesiveness it has had over decades , there was a time when every digital device you touched was running a java binary in the background. 

Reviewer: ```suggestion
For most of my career I have been a Java developer, I fell in love with it way back in college, hell I even gatekeeped teaching it when I was a university lecturer (a story for another day). I understood and still believe in what Java stands for and the stability and confidence it brings, not forgetting the pervasiveness it has had over decades, there was a time when every digital device you touched was running a Java binary in the background. 
```

😂 I can't wait to hear this story!

Now with over a decade of this Java knowledge and use sprinkled with a bit of C# and a lot of Python , I have decided to learn and start building with Go .
***Why ...?*** One may be forced to wonder, I wil tell you a few of the reasons here .
1. The thrill of learning , I love learning new things , I love learning things that I can do with my hands, I have found myself learning to knit, I have learnt how to play a guitar I have learnt how to play a mbira( and can claim to be pretty good at it) , I even have a harmonica on the way from from Germany <mark>(Hohner Special 20 in C)<mark>.

> 2. Now the bread and butter reason, I recently moved into a role where I primarily interact with a Go codebase, Every now and then I am called to sift through the code to understand and intepret the logic, While I can safely do this without much knowledge of go, just using my knowledge of programming I can read through and figure it out, But now it's a new challenge that makes me want to know more about it and figure out why we have high value products written completely in go, what's the pull factor. 

Reviewer: ```suggestion
2. Now the bread and butter reason, I recently moved into a role where I primarily interact with a Go codebase, and every now and then I am called to sift through the code to understand and interpret the logic. While I can safely do this without much knowledge of Go, just using my knowledge of programming I can read through and figure it out, but now it's a new challenge that makes me want to know more about it and figure out why we have high value products written completely in Go, what's the pull factor? 
```

> 3. I found a mentor who is an expert in go and in my selfish thinking I cannot not use this opportunity to learn. 

Reviewer: ```suggestion
3. I found a mentor who is an expert in Go and in my selfish thinking I cannot not use this opportunity to learn. 
```
Great initiative 😁

Author: keyword "selfish" 😁

So now that we have got that out of the way , the question is going to be How am I going to learn Go programming. 
The process:
I am going to use a few techniques and resources at my disposal to learn go as fast as I can, considering I actually don't have a lot of time to spare efficiency is going to be the one very important tool in the bag. 

### Here are the resources I have at my disposal .

> * I have this [Learning go with Tests ](https://quii.gitbook.io/learn-go-with-tests) , Very beginner friendly and easy to use tutorial , I like it not because its beginner friendly but rather It follows tdd religiously. I love **TDD** but in work situations I have rarely fully implemented **TDD** . I have mostly used it or was compelled to use when working with an existing codebase, one i did not write or I wrote but its now legacy . For my own cofidence I would write the tests first just to make sure I don't bring a wrecking ball and destroy the whole thing , So I am using this as an opportunity to get back to **TDD** and get it ingrained in me.

Reviewer: ```suggestion
* I have this [Learn Go with Tests ](https://quii.gitbook.io/learn-go-with-tests), a very beginner friendly and easy to use tutorial. I like it not because it's beginner friendly but rather because it follows TDD religiously. I love **TDD** but in work situations I have rarely fully implemented **TDD** . I have mostly used it or was compelled to use when working with an existing codebase, one I did not write or I wrote but it's now legacy . For my own confidence I would write the tests first just to make sure I don't bring a wrecking ball and destroy the whole thing, so I am using this as an opportunity to get back to **TDD** and get it ingrained in me.
```

* I already mentioned my mentor whom I meet with regulary, I can bring as many questions as I desire to him and we talk through the deep technical concepts in Go thinking. 

> * Again as I already mentioned I have codebases in Go, where i can immediately see how the concepts are being implemented . 

Reviewer: ```suggestion
* I already mentioned my mentor whom I meet with regularly, I can bring as many questions as I desire to him and we talk through the deep technical concepts in Go thinking. 
* Again as I already mentioned I have codebases in Go, where I can immediately see how the concepts are being implemented . 
```
